source/opt/trim_capabilities_pass.h

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_OPT_TRIM_CAPABILITIES_PASS_H_
+#define SOURCE_OPT_TRIM_CAPABILITIES_PASS_H_
+
+#include <algorithm>
+#include <functional>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "source/opt/ir_context.h"
+#include "source/opt/module.h"
+#include "source/opt/pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// This is required for NDK build. The unordered_set/unordered_map
+// implementation don't work with class enums.
+struct UnsignedHash {
+  unsigned operator()(spv::Capability value) const {
+    return static_cast<unsigned>(value);
+  }
+
+  unsigned operator()(spv::Op value) const {
+    return static_cast<unsigned>(value);
+  }
+};
+using CapabilitySet = std::unordered_set<spv::Capability, UnsignedHash>;
+
+// All the capabilities supported by this optimization pass. If your module
+// contains unsupported instruction, the pass could yield bad results.
+static const CapabilitySet kSupportedCapabilities{
+    spv::Capability::Linkage,
+    spv::Capability::Shader,
+    spv::Capability::ShaderClockKHR,
+};
+
+// Those capabilities disable all transformation of the module.
+static const CapabilitySet kFordiddenCapabilities{
+    spv::Capability::Linkage,
+};
+
+// Those capabilities are never removed from a module because we cannot
+// guess from the SPIR-V only if they are required or not.
+static const CapabilitySet kUntouchableCapabilities{
+    spv::Capability::Shader,
+};
+
+static const std::unordered_map<spv::Capability, std::string, UnsignedHash>
```

Comment:
Fixed the trivially constructible/destructible part.

For the featureManager, I am not using it because it doesn't remove instructions from the module. The FeatureManager does have a Remove() function, but it only removes it from the set, but doesn't apply the changes to the underlying module.
Probably something we should fix/improve in a refactorization?
Because of this weird thing, I decided to keep a lower-level interaction with instructions directly.

For the extensions/capability mapping, do you mean I should add a function to, given a list of capabilities, get the list of required extensions, and this depending on the spirv version?

If you think this could be useful outside of this pass, sure, we can add that. Subsequent PR or this PR?